File: jupyter/Qtjupyter/python/module/derivative.py
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from typing import Callable

from module.asset import Asset


# General Derivative class
class Derivative:
    def __init__(self,underlying:Asset,structure: Callable[[float], float],price:float,strike:float=None,name:str="None"):
        self.underlying = underlying
        self.f = structure
        self.name=name
        self.price=price
        self.strike=strike
    def plotReturn(self):
        t0= self.uValue
        size = np.abs(t0/10)
        t = np.arange(t0-size,t0+size, size/100)
        #add Axis
        plt.axhline(0, color='black')
        plt.axvline(self.uValue, color='black')
        #add Graph
        plt.plot(t,self.f(t)-self.price,"r")
        plt.plot(self.uValue,self.f(self.uValue)-self.price,"gx")
        
        plt.show()
    def printProps(self,withPlot=True):
        otp=f"""
        {'-'*40}
        Asset: {self.underlying.id}
        Derivative type: {self.name}
        Underlying value: {self.uValue}
        Derivative value: {self.value}
        Price paid: {self.price}
        Strike price: {self.strike}
        Absolute return: {self.absReturn}
        Percentage return: {self.pctReturn}
        Log return: {self.logReturn}
        {'-'*40}
        """
        print(otp)
        if withPlot:
            self.plotReturn()
        
    
    @property
    def uValue(self)-> float:
        return self.underlying.data.Close[-1]
    @property
    def value(self) -> float:
        return self.f(self.uValue)
    @property
    def absReturn(self) -> float:
        return self.f(self.uValue)-self.price
    @property
    def pctReturn(self) -> float:
        return self.absReturn/self.price
    @property
    def logReturn(self) -> float:
        if self.value<=0:
            return np.NINF
        return np.log(self.value/self.price)
    
#Special classes

class Buy(Derivative):
    def __init__(self,underlying:Asset,premium:float=0):
        pricePaid=underlying.data.Close[-1] + premium
        f = lambda x: x
        super().__init__(underlying,f,pricePaid,name="Buy")
        
class Sell(Derivative):
    def __init__(self,underlying:Asset,premium:float=0):
        pricePaid=-underlying.data.Close[-1] + premium
        f = lambda x: -x
        super().__init__(underlying,f,pricePaid,name="Sell")

class LongCall(Derivative):
    def __init__(self,underlying:Asset,strike:float,premium:float=0):
        pricePaid = premium
        f = lambda x : np.maximum(x-strike,0)
        super().__init__(underlying,f,pricePaid,strike,name="Long Call")

class LongPut(Derivative):
    def __init__(self,underlying:Asset,strike:float,premium:float=0):
        pricePaid = premium
        f = lambda x : np.maximum(-x+strike,0)
        super().__init__(underlying,f,pricePaid,strike,name="Long Put")

class ShortCall(Derivative):
    def __init__(self,underlying:Asset,strike:float,premium:float=0):
        pricePaid = -premium
        f = lambda x : np.minimum(-x+strike,0)
        super().__init__(underlying,f,pricePaid,strike,name="Short Call")
        
class ShortPut(Derivative):
    def __init__(self,underlying:Asset,strike:float,premium:float=0):
        pricePaid = -premium
        f = lambda x : np.minimum(+x-strike,0)
        super().__init__(underlying,f,pricePaid,strike,name="Short Pull")
        

File: src/Derivative.jl

export 
    Derivative,
    Buy,
    Sell,
    LongCall,
    LongPut,
    ShortCall,
    ShortPut,
    plot,
    printProps,
    uValue,
    value,
    absReturn,
    pctReturn,
    logReturn




"""
    Derivative

Abstract type for all derivatives.
"""
abstract type Derivative end

Base.show(io::IO,D::Derivative) = print(io,"Derivative of Type '$(D.name)' on $(D.underlying.ticker)")

function uValue(D::Derivative)
    return value(D.underlying)
end
function value(D::Derivative)
    return D.structure(uValue(D))
end
function absReturn(D::Derivative)
    return D.structure(uValue(D)-D.price)
end
function pctReturn(D::Derivative)
    return absReturn(D)/D.price
end
function logReturn(D::Derivative)
    value(D)<=0 ? -Inf : log(value(D)/D.price);
end


"""
    plot(D::<Derivative)

Plots the payoffstructure.

# Examples
```jldoctest
Random.seed!(456);
A=Asset();
LC=LongCall(A,value(A),10);
plot(LC)
# output

                       KPGR, Long Call              
         ┌────────────────────────────────────────┐ 
      40 │                                        │ 
         │                                       .│ 
         │                                      .'│ 
         │                                     :' │ 
         │                                   .'   │ 
         │                                 .:     │ 
         │                                .'      │ 
   y     │                              .:'       │ 
         │                             .'         │ 
         │                            :'          │ 
         │                          .'            │ 
         │                         :              │ 
         │''''''''''''''''''''''':''''''''''''''''│ 
         │                     .:'                │ 
     -10 │.....................'                  │ 
         └────────────────────────────────────────┘ 
          44                                   133  
                              x                     

```

"""
function plot(D::Derivative)
    f=D.structure
    v=uValue(D)
    x=range(v-v/2,v+v/2)
    lineplot(
        x,
        f.(x).-D.price,
        xlim= floor.(extrema(x)),
        title=D.underlying.ticker*", "*D.name,
        canvas=DotCanvas,
        xlabel="x",
        ylabel="y")
end
 
"""
    printProps(D::Derivative)

Print all the current properties of the input derivative. 

# Examples

```jldoctest
Random.seed!(456);
A = Asset();
B = Buy(A,10);
printProps(B)
# output

========================================
Assets: KPGR
Derivative type: Buy
Underlying value: 100.16749318215984
Derivative value: 100.16749318215984
Price paid: 110.16749318215984
Strike price: None
Absolute return: -10.0
Percentage return: -0.09077087724475305
Log return: -0.09515815632970726
========================================
```

"""
function printProps(D::Derivative)
    if hasproperty(D,:strike) 
        str = D.strike
    else
        str = "None"
    end
    otp ="="^40*"\n"*"""
    Assets: $(D.underlying.ticker)
    Derivative type: $(D.name)
    Underlying value: $(uValue(D))
    Derivative value: $(value(D))
    Price paid: $(D.price)
    Strike price: $(str)
    Absolute return: $(absReturn(D))
    Percentage return: $(pctReturn(D))
    Log return: $(logReturn(D))
    """*"="^40
    print(otp)
end

"""


"""
mutable struct Buy <: Derivative
    underlying::Asset
    structure::Function
    name::String
    price::Number
    function Buy(underlying::Asset,premium::Number=0)
        this = new()
        this.underlying=underlying
        this.structure = x->x
        this.name = "Buy"
        this.price = value(underlying) + premium 
        return this
    end
end



mutable struct Sell <: Derivative
    underlying::Asset
    structure::Function
    name::String
    price::Number
    function Sell(underlying::Asset,premium::Number=0)
        this = new()
        this.underlying=underlying
        this.structure = x->-x
        this.name = "Sell"
        this.price = -value(underlying) + premium 
        return this
    end
end

mutable struct LongCall <: Derivative
    underlying::Asset
    structure::Function
    name::String
    price::Number
    strike::Number
    function LongCall(underlying::Asset,strike::Number,premium::Number=0)
        this = new()
        this.underlying=underlying
        this.structure = x-> max(x-strike,0)
        this.name = "Long Call"
        this.price = premium 
        this.strike = strike
        return this
    end
end

mutable struct LongPut <: Derivative
    underlying::Asset
    structure::Function
    name::String
    price::Number
    strike::Number
    function LongPut(underlying::Asset,strike::Number,premium::Number=0)
        this = new()
        this.underlying=underlying
        this.structure = x-> max(-x+strike,0)
        this.name = "Long Put"
        this.price = premium 
        this.strike = strike
        return this
    end
end

mutable struct ShortCall <: Derivative
    underlying::Asset
    structure::Function
    name::String
    price::Number
    strike::Number
    function ShortCall(underlying::Asset,strike::Number,premium::Number=0)
        this = new()
        this.underlying=underlying
        this.structure = x-> min(-x+strike,0)
        this.name = "Short Call"
        this.price = -premium 
        this.strike = strike
        return this
    end
end

mutable struct ShortPut <: Derivative
    underlying::Asset
    structure::Function
    name::String
    price::Number
    strike::Number
    function ShortPut(underlying::Asset,strike::Number,premium::Number=0)
        this = new()
        this.underlying=underlying
        this.structure = x-> min(x-strike,0)
        this.name = "Short Put"
        this.price = -premium 
        this.strike = strike
        return this
    end
end

#SP=ShortPut(x,100,10)
#printProps(SP)
#B=Buy(x)
#printProps(B)
#plot(SP)
